recipes/catch2/2.x.x/conanfile.py

```diff
@@ -51,79 +42,80 @@ def config_options(self):
 
     def configure(self):
         if not self.options.with_main:
-            del self.options.fPIC
-            del self.options.with_benchmark
+            try:
+                del self.options.fPIC
+            except Exception:
+                pass
+            try:
+                del self.options.with_benchmark
+            except Exception:
+                pass
 
     def package_id(self):
         if not self.options.with_main:
             self.info.header_only()
 
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
     def validate(self):
-        if tools.Version(self.version) < "2.13.1" and self.settings.arch == "armv8":
+        if Version(self.version) < "2.13.1" and self.settings.arch == "armv8":
             raise ConanInvalidConfiguration("ARMv8 is supported by 2.13.1+ only! give up!")
-        if self.options.with_main and tools.Version(self.version) < "2.13.4":
+        if self.options.with_main and Version(self.version) < "2.13.4":
             raise ConanInvalidConfiguration("Option with_main not supported with versions < 2.13.4")
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
-
-    @functools.lru_cache(1)
-    def _configure_cmake(self):
-        cmake = CMake(self)
-        cmake.definitions["BUILD_TESTING"] = "OFF"
-        cmake.definitions["CATCH_INSTALL_DOCS"] = "OFF"
-        cmake.definitions["CATCH_INSTALL_HELPERS"] = "ON"
-        cmake.definitions["CATCH_BUILD_STATIC_LIBRARY"] = self.options.with_main
-        cmake.definitions["enable_benchmark"] = self.options.get_safe("with_benchmark", False)
-        cmake.definitions["CATCH_CONFIG_PREFIX_ALL"] = self.options.with_prefix
+        get(self, **self.conan_data["sources"][self.version], destination=self.source_folder, strip_root=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["BUILD_TESTING"] = False
+        tc.cache_variables["CATCH_INSTALL_DOCS"] = False
+        tc.cache_variables["CATCH_INSTALL_HELPERS"] = "ON"
+        tc.cache_variables["CATCH_BUILD_STATIC_LIBRARY"] = self.options.with_main   # FIXME cache_variables or just variables?
```

Comment:
In case of doubt: https://github.com/conan-io/conan/issues/11937#issuecomment-1224053745

Plus: 
- https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmaketoolchain.html#cache-variables
- https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmaketoolchain.html#variables

There are two main cases to use `cache_variables`:

- When a definition is configured **before** `project()`
- When an option uses `CACHE`. Need to check project's CMakeLists.txt

Reply:
The docs currently focus on the difference that CACHE is for "single config" compare to "multiple config",
and says `for the majority of cases cache_variables is what you probably want to use` ... so this seems to be a complex topic...
I see almost all recipes use variables, but it seems like if you really want the variable to be set, you should use cache_variables and there is less room for surprises?

Reply:
In CMake, cache variables are how CMake and projects expose options to be set by consumers. Cache variables are meant to be able to be preemptible by consumers. If a consumer sets it before a project does, the project won't change the value. I think you're right, @paulharris, that we should prefer setting cache variables instead when possible.

There are of course, tons of CMake projects that don't respect cache variables and set normal variables with the same name which causes all sorts of problems.... ugh

Reply:
Hopefully improvements for this are coming with a future CMake?

Reply:
https://github.com/conan-io/conan-center-index/pull/13971 not the best but it's a start